### PR TITLE
Fix landed and auto-disarm hysteresis

### LINKED ISSUE #5083: Auto-disarm in RTL reported
There seems to be a scary bug where RTL caused all motors to stop immediately in flight.

More details in https://groups.google.com/forum/#!topic/px4users/UxPZQ6G4KVs.

### PR DESCRIPTION
**Pull request by julianoes**
This is an attempt to fix #5083.

The issue showed two cases where hysteresis was used wrongly. Both times, the timestamp was not updated correctly or consistently. I therefore went ahead and created a class which abstracts hysteresis away and replaced the two cases.

Please review and cross-test, thanks.

### CODE REVIEW

**Comment by bkueng** [comment]
looks good! can you comment what the Hysteresis class members do (arguments, return values)?

2845ce6109d2bdf7 is not really necessary, as it's time differences and it only needs more memory

**Comment by dagar** [comment]
I've had a few FW flights where land was briefly registered in air causing the home to reset and then failed auto landings. Could we pull update() into the base LandDetector including the hysteresis?

+1 for new code with testing.

**Comment by julianoes** [comment]
> Could we pull update() into the base LandDetector including the hysteresis?

Good point, we should adapt the fixedwing code as well. I'll do it.

**Comment by julianoes** [comment]
> 2845ce6 is not really necessary, as it's time differences and it only needs more memory

@bkueng agreed, I just wasn't sure if it was gonna bite us some day.

**Comment by julianoes** [comment]
@dagar: your request to re-use the update function across the 3 estimators made perfect sense but it lead to some bigger refactor along the way :).

Any chance you could do a review of the refactored code?

**Comment by julianoes** [comment]
Jenkins: test this please

File: src/modules/land_detector/LandDetector.h [hunk 1]
```diff
@@ -31,101 +31,131 @@
...
-enum LandDetectionResult {
-	LANDDETECTION_RES_FLYING = 0,	/**< UAV is flying */
-	LANDDETECTION_RES_LANDED = 1,	/**< Land has been detected */
-	LANDDETECTION_RES_FREEFALL = 2	/**< Free-fall has been detected */
-};
 
 class LandDetector
 {
 public:
+	enum class LandDetectionState {
```
Comment (dagar) [inline src/modules/land_detector/LandDetector.h]: Strongly typed enums!

**Comment by dagar** [comment]
Big diffs are hard to review and I wasn't familiar with the existing code, but it generally looks good. The common topics could live in the base, but it doesn't really matter.

For now we should keep this PR focused on fixing that nasty disarm in air bug, but I can't help but wonder if more of the land detector should be shared between FW and MC. The base could check xy, z velocity, arming, etc. FW would add airspeed checks, MC would have rotating. Right now MC has some nicer features that capture user intent like checking manual.z if in a manual mode which also makes sense for FW.
https://github.com/PX4/Firmware/blob/c86c1d907c24261a3d0a98d7f80cb1e8d04e203c/src/modules/land_detector/MulticopterLandDetector.cpp#L175

Thanks for doing the general solution. I'm doing some testing this week so I can add this to the list.

**Comment by julianoes** [comment]
@dagar Thanks for the review. I agree that we could probably improve the detectors but let's keep it at this fix and refactor for now and do fancier changes in a next iteration.

Testing would be appreciated, I'm also planning some this week.

**Comment by julianoes** [comment]
@CarlOlsson This is one of the PRs I mentioned today that I would appreciate if you could do some testing, of course with something cheap that is allowed to fall out of the sky :).

**Comment by CarlOlsson** [comment]
Hi, actually I don't have access to a cheap little quad at the moment. Would be nice but it just never got prioritized.

File: src/modules/land_detector/FixedwingLandDetector.cpp [hunk 2]
```diff
@@ -31,22 +31,23 @@
  *
  ****************************************************************************/
 
-/**
+/*
```
Comment (LorenzMeier) [inline src/modules/land_detector/FixedwingLandDetector.cpp]: It needs two starts to be considered a Doxygen comment. Please add the star back in.

Comment (julianoes) [inline src/modules/land_detector/FixedwingLandDetector.cpp]: Ah :) will do.

**Comment by dagar** [comment]
FYI - with unit tests the goal was to have the same set of tests that runs on nuttx, sitl, snappy, etc. I tried pretty hard to get gtest working on nuttx, but it just wasn't happening.

**Comment by LorenzMeier** [comment]
@julianoes Can you rebase this on master?

**Comment by julianoes** [comment]
Done, this state is what I'm planning to test today.

**Comment by LorenzMeier** [comment]
Because master is not necessarily ready to fly and because this issue is causing crashes I'm going to put it on master in any case and we'll hotfix if anything is remaining.

**Comment by julianoes** [comment]
Also, I have flown this successfully several times today on a Pixfalcon.

**Pull request merged**